streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java

```diff
@@ -439,6 +439,87 @@ public void shouldPerformSelectKeyWithRepartitionOperation(final String topology
         assertEquals(1, countOccurrencesInTopology(topology, "Sink: .*-repartition.*"));
     }
 
+    @ParameterizedTest
+    @ValueSource(strings = {StreamsConfig.OPTIMIZE, StreamsConfig.NO_OPTIMIZATION})
+    public void shouldNotRepartitionWithMarkAsPartitionedFollowingSelectKey(final String topologyOptimization) throws Exception {
+        final long timestamp = System.currentTimeMillis();
+
+        sendEvents(
+                timestamp,
+                Arrays.asList(
+                        new KeyValue<>(1, "10"),
+                        new KeyValue<>(2, "20")
+                )
+        );
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()))
+                .selectKey((key, value) -> Integer.valueOf(value))
+                .markAsPartitioned(Named.as("partition-preserved"))
+                .groupByKey()
+                .count()
+                .toStream()
+                .to(outputTopic);
+
+
+        startStreams(builder, createStreamsConfig(topologyOptimization));
+
+        validateReceivedMessages(
+                new IntegerDeserializer(),
+                new LongDeserializer(),
+                Arrays.asList(
+                        new KeyValue<>(10, 1L),
+                        new KeyValue<>(20, 1L)
+                )
+        );
+
+        final String topology = builder.build().describe().toString();
+
+        assertEquals(0, countOccurrencesInTopology(topology, "Sink: .*-repartition.*"));
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {StreamsConfig.OPTIMIZE, StreamsConfig.NO_OPTIMIZATION})
+    public void shouldNotRepartitionWithMarkAsPartitionedFollowingMap(final String topologyOptimization) throws Exception {
+        final String topicBMapperName = "topic-b-mapper";
+        final long timestamp = System.currentTimeMillis();
+
+        sendEvents(
+                timestamp,
+                Arrays.asList(
+                        new KeyValue<>(1, "10"),
+                        new KeyValue<>(2, "20")
+                )
+        );
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()))
+                .map(KeyValue::new, Named.as(topicBMapperName))
```

Comment:
This seems to be pretty much the same as the test above? Do we gain much adding it with regard to "useful" test coverage?

It might be more interesting to add a "branching" case, as well as a "downstream" key-changing operation:
```
// branching
KStream s = builder.stream(...).map(...)

s.markAsPartitioned(Named.as("partition-preserved"))
 .groupByKey()
 .count()
 .toStream()
 .to(outputTopic);

// re-using `s` here, "branches / fans-out" the topology, and because we call `groupByKey()` directly on `s` we should get a repartition topic here (we don't see the `markAsPartitions()` operation from above, and it should not impact `s` itself.
s.groupByKey()
 .count()
 .toStream()
 .to(outputTopic2);
 
 
// downstream map
KStream s = builder.stream(...)
    .map(...)
    .markAsPartitioned(Named.as("partition-preserved"))
    .map(...) // insert a second map -- this should make the upstream `markAsPartitioned()` void and we should get a repartition topic again
    .groupByKey()
    .count()
    .toStream()
    .to(outputTopic);
    
// maybe also: branch after `markAsPartitions`
KStream s = builder.stream(...)
    .map(...)
    .markAsPartitioned(Named.as("partition-preserved"));

// map() after "markAsPartitios in this branch -- should trigger repartitoning
s.map(...)
 .groupByKey()
 .count()
 .toStream()
 .to(outputTopic);

// no reparation topic in this branch
s.groupByKey()
 .count()
 .toStream()
 .to(outputTopic2);
```